### ISSUE TITLE
Calls are recorded after applying the best rule, not when received

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#1583]
From StackOverflow question [FakeItEasy assert that method call to method](https://stackoverflow.com/questions/55675942/fakeiteasy-assert-that-method-call-to-method/55677429#55677429).

@baruchiro had a test that passed in 2.2.0 and failed in 5.1.0:

```c#
public class Foo
{
    public virtual void MethodA()
    {
        MethodB();
    }

    public virtual void MethodB() { }
}

…

var foo_fake = A.Fake<Foo>(options => options.CallsBaseMethods());

foo_fake.MethodA();

A.CallTo(() => foo_fake.MethodA()).MustHaveHappened()
    .Then(A.CallTo(() => foo_fake.MethodB()).MustHaveHappened());
```

The failure indicated that both calls happened, but in the wrong order.

We inadvertently changed the behaviour in #1124, released in 3.4.2, when we started recording the call in the FakeManager after applying the best-matching rule, rather than in the `CastleInvocationCallAdapter`.

We could record the call before looking for the best rule to apply. I think this is a good fix anyhow, as it more accurately reflects when the call was made, irrespective of any time FakeItEasy takes to figure out what to do with it.

**Comment by blairconrad** [comment]
StackOverflow user Nikosi, whom I believe to be @nkosihenry (sorry to both of you if I'm wrong), suggested some options, which I respond to now:

> You could consider making it configurable depending on the tastes of the user. the default could be after completion with the option to change it to before. But then again it might end up becoming scope creep.

I agree. I think it's scope creep, and better to record when the call is intercepted.

> What do you think about MustHaveStarted. A call to MethodA MustHaveStarted then a call to MethodB...

I think it's a good name, but prefer to keep the current name and just change when we record the call, restoring it to the original behaviour.

Of course, rebuttals are welcome.

**Comment by blairconrad** [comment]
In case we do want to make the change, I've created #1584.

**Comment by nkosihenry** [comment]
@blairconrad, you got the name correct. 

I mentioned `MustHaveStarted` as its own check and not to replace/change the `MustHaveHapppened`.

For me the semantics are in the working. Happened would be checked after invocation has completed. while `Started` would be checked before completion.

**Comment by blairconrad** [comment]
@nkosihenry, thanks. I see your point about the wording and the semantics. Sadly, to some degree, this is in the eye of the reader.

To me, "A call to blah must have happened" talks more about when the call happened.
I'm quick to say "no" to features, but I don't really see much benefit in differentiating between when a call is begun and when it finishes, especially when the goings-on between the two events should all be FakeItEasy or user-supplied test code.
In the interest of keeping the API small, I think we should supply only the one check, and I think `MustHaveHappened` adequately conveys the intent (also requires no rework for the vast majority of users for whom

**Comment by nkosihenry** [comment]
Agreed. Keeping it Small and Simple (KISS). `MustHaveHappened` works for what it was intended.

**Comment by thomaslevesque** [comment]
Most people probably don't need the distinction between when the call has started and when it has completed. At least, it never came up in all the years FakeItEasy has been in use. So +1 for keeping just `MustHaveHappened`.

Regarding the issue itself, I agree that we should record the call before applying the rule, if it doesn't break anything.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 5.1.1](https://github.com/FakeItEasy/FakeItEasy/releases/tag/5.1.1).
